core/primitives/src/views.rs

```diff
@@ -1069,23 +1107,115 @@ impl From<FinalExecutionOutcomeWithReceiptView> for FinalExecutionOutcomeView {
     }
 }
 
-#[derive(BorshSerialize, BorshDeserialize, Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
-pub struct ValidatorStakeView {
-    pub account_id: AccountId,
-    pub public_key: PublicKey,
-    #[serde(with = "u128_dec_format")]
-    pub stake: Balance,
-}
+#[cfg(feature = "protocol_feature_block_header_v3")]
+pub mod validator_stake_view {
+    use crate::serialize::u128_dec_format;
+    use crate::types::validator_stake::ValidatorStake;
+    use borsh::{BorshDeserialize, BorshSerialize};
+    use near_crypto::PublicKey;
+    use near_primitives_core::types::{AccountId, Balance};
+    use serde::{Deserialize, Serialize};
+
+    #[derive(
+        BorshSerialize, BorshDeserialize, Serialize, Deserialize, Debug, Clone, Eq, PartialEq,
+    )]
+    #[serde(tag = "version")]
+    pub enum ValidatorStakeView {
```

Comment:
You can simply add one more field to `ValidatorStakeView`. In terms of json serialization it will be backward compatible

Reply:
Yes, that's right. The `#[serde(tag = "version")]` attribute makes it serialize the enum variant as an additional field (e.g. `"version": "v1"`) instead of as nested objects. I think this is useful because it means clients can act differently depending on the version of the structure (during the transition period `version` not being a present field will also be relevant). For example, in our [lightclient pytest](https://github.com/near/nearcore/blob/master/pytest/lib/lightclient.py#L145) we validate the `next_bp_hash` field, and this can only be done properly if the version is present in the json structure (because the version -- or lack of version -- influences the borsh serialization). I actually need to fix that in my PR here (thanks for reminding me!).

The reason to make `ValidatorStakeView` an enum at all is because it also implements borsh serialization, where adding an extra field is not backwards compatible. Presently it appears to not be used, so I didn't worry about backwards compatibility in the borsh serialization when introducing the enum in this PR, but it looks like it will be used for [syncing](https://github.com/near/nearcore/blob/master/core/primitives/src/syncing.rs#L217) in the future and so I expect we will be happy it is an enum down the road.

Reply:
You can see me using `version` in the json serialization now: https://github.com/near/nearcore/pull/4115/commits/8256535d11bf08a803fa5462fbf4f870fe6beb10

The lightclient python test now passes for both stable and nightly binaries.

Reply:
> but it looks like it will be used for syncing in the future and so I expect we will be happy it is an enum down the road.

I don't think syncing will use this struct. `view` struct should only be used in rpc responses.

Reply:
@frol what do you think?

Reply:
> I don't think syncing will use this struct. `view` struct should only be used in rpc responses.

The skeleton for using it is already in place. `EpochSyncResponse` is in `near_primitives::syncing` and uses `LightClientBlockView`, which contains `ValidatorStakeView`. Is it expected this will change?

Reply:
That is unfortunate. We should address that issue instead of perpetuating it.

Reply:
Alright, so it does not need to be an enum for syncing reasons. What do you think about my usage of the version field in the lightclient test? The test attempts to manually verify a hash so it needs to know the details of serialization (of the real ValidatorStake, not just the view), and hence the version is necessary information.

Reply:
You can achieve the same by checking whether the new field exists right?

Reply:
> You can achieve the same by checking whether the new field exists right?

That's true, though it seems a little hacky to me that someone would need to know what fields were introduced in which version to figure out what the correct serialization is. But then again, it's a little hacky to try and ad-hoc reconstruct the borsh serialization in the first place.

Let's see what @frol has to say about whether or not a `version` field is useful in the view struct.

Reply:
I believe it is a great foundation to have everything versioned. Having the internally tagged enum representation plays fine for the backward compatibility :+1:

My only ask would be to use a more explicit tag name: `validator_stake_struct_version` (I am not super happy about the naming, but plain `version` is super confusing, so if you have a better idea, go for it).

Reply:
Name changed to `validator_stake_struct_version` in https://github.com/near/nearcore/pull/4115/commits/b288f74bf34b0a8aa50da75ee9a3698179259623